### ISSUE TITLE
after_update running inappropriately (when no changes occurred)

### ISSUE THREAD

**Issue by zmoazeni** [rails/rails#26270]
Separated into a new issue from https://github.com/rails/rails/issues/13360#issuecomment-233399289

With help from @yukideluxe
### Steps to reproduce

``` ruby
gem 'activerecord', '5.0.0'
require 'active_record'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')

ActiveRecord::Schema.define(version: 20160718165350) do
  create_table "accounts", force: :cascade do |t|
    t.integer  "bank_id"
    t.integer  "person_id"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  add_index "accounts", ["bank_id"], name: "index_accounts_on_bank_id"
  add_index "accounts", ["person_id"], name: "index_accounts_on_person_id"

  create_table "banks", force: :cascade do |t|
    t.string   "name"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "people", force: :cascade do |t|
    t.string   "name"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end
end

class Person < ActiveRecord::Base
end

class Bank < ActiveRecord::Base
  has_many :accounts
  has_many :people, through: :accounts
end

class Account < ActiveRecord::Base
  belongs_to :bank
  belongs_to :person

  after_update :this_should_not_run

  def this_should_not_run
    raise "when did I ever update an account?"
  end
end

person = Person.create!(name: "William")
Bank.create!(name: "Stacks of Money", person_ids: [person.id])
```
### Expected behavior

`Account` after_update callbacks should never run, because the account was never updated, only created.
### Actual behavior

`Account` after_update runs with this stacktrace:

```
∴ bundle exec ruby callback_explosion.rb
-- create_table("accounts", {:force=>:cascade})
   -> 0.0165s
-- add_index("accounts", ["bank_id"], {:name=>"index_accounts_on_bank_id"})
   -> 0.0003s
-- add_index("accounts", ["person_id"], {:name=>"index_accounts_on_person_id"})
   -> 0.0004s
-- create_table("banks", {:force=>:cascade})
   -> 0.0003s
-- create_table("people", {:force=>:cascade})
   -> 0.0003s
-- initialize_schema_migrations_table()
   -> 0.0033s
callback_explosion.rb:46:in `this_should_not_run': when did I ever update an account? (RuntimeError)
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:382:in `block in make_lambda'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:207:in `block in halting_and_conditional'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:456:in `block in call'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:456:in `each'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:456:in `call'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:101:in `__run_callbacks__'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:750:in `_run_update_callbacks'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/callbacks.rb:306:in `_update_record'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/timestamp.rb:81:in `_update_record'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/persistence.rb:534:in `create_or_update'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/callbacks.rb:298:in `block in create_or_update'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:126:in `call'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:506:in `block (2 levels) in compile'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:455:in `call'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:101:in `__run_callbacks__'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:750:in `_run_save_callbacks'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/callbacks.rb:298:in `create_or_update'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/persistence.rb:152:in `save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/validations.rb:50:in `save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/attribute_methods/dirty.rb:30:in `save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/transactions.rb:324:in `block in save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/transactions.rb:395:in `block in with_transaction_returning_status'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract/database_statements.rb:230:in `transaction'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/transactions.rb:211:in `transaction'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/transactions.rb:392:in `with_transaction_returning_status'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/transactions.rb:324:in `save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/suppressor.rb:45:in `save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/associations/has_many_through_association.rb:89:in `save_through_record'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/associations/has_many_through_association.rb:47:in `insert_record'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/autosave_association.rb:400:in `block in save_collection_association'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/autosave_association.rb:391:in `each'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/autosave_association.rb:391:in `save_collection_association'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/autosave_association.rb:185:in `block in add_autosave_association_callbacks'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/autosave_association.rb:158:in `instance_eval'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/autosave_association.rb:158:in `block in define_non_cyclic_method'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:382:in `block in make_lambda'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:207:in `block in halting_and_conditional'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:456:in `block in call'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:456:in `each'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:456:in `call'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:101:in `__run_callbacks__'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:750:in `_run_create_callbacks'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/callbacks.rb:302:in `_create_record'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/timestamp.rb:68:in `_create_record'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/persistence.rb:534:in `create_or_update'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/callbacks.rb:298:in `block in create_or_update'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:126:in `call'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:506:in `block (2 levels) in compile'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:455:in `call'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:101:in `__run_callbacks__'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb:750:in `_run_save_callbacks'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/callbacks.rb:298:in `create_or_update'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/persistence.rb:152:in `save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/validations.rb:50:in `save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/attribute_methods/dirty.rb:30:in `save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/transactions.rb:324:in `block in save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/transactions.rb:395:in `block in with_transaction_returning_status'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `block in transaction'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract/transaction.rb:189:in `within_new_transaction'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `transaction'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/transactions.rb:211:in `transaction'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/transactions.rb:392:in `with_transaction_returning_status'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/transactions.rb:324:in `save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/suppressor.rb:45:in `save!'
  from /Users/zmoazeni/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/persistence.rb:51:in `create!'
  from callback_explosion.rb:51:in `<main>'
```

Additionally the changes and previous_changes are unexpected for an update callback:

``` ruby
# changes
{}

# previous_changes
{"id"=>[nil, 1], "bank_id"=>[nil, 1], "person_id"=>[nil, 1], "created_at"=>[nil, 2016-07-18 17:41:17 UTC], "updated_at"=>[nil, 2016-07-18 17:41:17 UTC]}
```
### System configuration

**Rails version**: tested this with versions 4.2.5.2, 4.2.6, 4.2.7, and 5.0.0.

**Ruby version**: 2.3.1, but should be unrelated

**Comment by prakashmurthy** [comment]
The above script for reproducing the behavior fails if the required version of `activerecord` gem is not installed locally. Modified it to install the gem at the start.

<details>
  <summary>Script to install required gems and reproduce the behavior</summary>
```ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  # Activate the gem you are reporting the issue against.
  # gem 'activerecord', "5.0.0" 
  gem 'activerecord', "4.2.7.1"
  # gem 'minitest'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table "accounts", force: :cascade do |t|
    t.integer  "bank_id"
    t.integer  "person_id"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  add_index "accounts", ["bank_id"], name: "index_accounts_on_bank_id"
  add_index "accounts", ["person_id"], name: "index_accounts_on_person_id"

  create_table "banks", force: :cascade do |t|
    t.string   "name"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "people", force: :cascade do |t|
    t.string   "name"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end
end

class Person < ActiveRecord::Base
end

class Bank < ActiveRecord::Base
  has_many :accounts
  has_many :people, through: :accounts
end

class Account < ActiveRecord::Base
  belongs_to :bank
  belongs_to :person

  after_update :this_should_not_run

  def this_should_not_run
    raise "when did I ever update an account?"
  end
end

class BugTest1 < Minitest::Test
  def test_create_person_and_bank
    person = Person.create!(name: "William")
    assert Bank.create!(name: "Stacks of Money", person_ids: [person.id])
  end
end
```
</details>

The script fails on 5.x and 4.x versions, but works with `3.2.22.4`.

**Comment by maclover7** [comment]
Off the top of my head, I think that an `Account` is being updated, because you are technically creating an `Account` behind the scenes by passing in `person_ids` to `Bank.create!`. If you remove the `raise`, and run through the script again, `Account.count` should have incremented by one 😬

**Comment by zmoazeni** [comment]
@maclover7 nah I disagree. It does create a record (no record exists prior and then one exists after) but it doesn't update a record (an existing row with an id). 

You can see the odd behavior when you look at the changes (id is said to change from nil to a number). 

after_update behaves fine for other models. For this case however after_update behaves more like after_save.

**Comment by kamipo** [comment]
This has been fixed since 5.1.
